### ISSUE TITLE
Shift in AST IDs affects inlining and expression splitting decisions in Yul Optimizer, causing bytecode differences

### ISSUE THREAD

**Issue by klkvr** [ethereum/solidity#14829]
## Description

This is probably related to #14494, I am using solc 0.8.23 and still encounter the issue

## Steps to Reproduce

Here are two standard JSON input files:
[input1.json](https://github.com/ethereum/solidity/files/14132480/input1.json)
[input2.json](https://github.com/ethereum/solidity/files/14132482/input2.json)

The only difference between those is a small source file:
```
"contracts/DummyContract.sol": {
    "content": "contract DummyContract {}"
},
```

We are interested in the `UniversalRouter` bytecode. It was obtained by using the following command:
```
cat input1.json | '/path/to/0.8.23' --standard-json | jq '.contracts."contracts/UniversalRouter.sol".UniversalRouter.evm.bytecode.object' > bytecode1.txt 
```

Resulted bytecodes:
[bytecode1.txt](https://github.com/ethereum/solidity/files/14132486/bytecode1.txt)
[bytecode2.txt](https://github.com/ethereum/solidity/files/14132487/bytecode2.txt)

Bytecodes are expected to be equal, however they have a pretty big diff:

<img width="1389" alt="image" src="https://github.com/ethereum/solidity/assets/62447812/362c21f7-f73f-4547-93eb-21137ac69040">

**Comment by cameel** [comment]
Thanks for the report! Unfortunately I can confirm that this is still reproducible on 0.8.24 and indeed looks like a difference in the bytecode coming from the optimizer. Including `DummyContract.sol` shifts AST IDs and they seem to have impact on the output even though they shouldn't.

It only happens with the `--optimize` flag. After inspecting the `--ir-optimized` output I see some IR differences that go beyond just slightly different names and immutable IDs. There seem to be two spots where decisions taken by the optimizer differ.

### IR differences
1. There's one spot where an expression does not get split into subexpressions:

    ```diff
                             let usr$firstWord := calldataload(expr_component_2)
                             let var_token0 := shr(96, usr$firstWord)
    -                        let _36 := 0xffffff
    -                        let _37 := add(expr_component_2, 23)
    -                        let _38 := calldataload(_37)
    -                        let var_token1 := shr(96, _38)
    -                        let expr_4 := fun_computePoolAddress(var_token0, var_token1, and(shr(72, usr$firstWord), _36))
    +                        let _36 := add(expr_component_2, 23)
    +                        let var_token1 := shr(96, calldataload(_36))
    +                        let expr_4 := fun_computePoolAddress(var_token0, var_token1, and(shr(72, usr$firstWord), 0xffffff))
    ```
2. There are two places where a function called `fun_swap()` gets inlined. Here's the first one, from external dispatch:

    ```diff
                                 default {
                                     if gt(23, expr_component_1) { revert(0, 0) }
    -                                let lengthOut := add(expr_component_1, not(22))
                                     if iszero(lt(expr_component_4, shl(255, 1))) { revert(0, 0) }
    -                                let _41 := negate_int256(expr_component_4)
    -                                if lt(lengthOut, 43)
    -                                {
    -                                    let _42 := mload(_2)
    -                                    mstore(_42, shl(224, 0x3b99b53d))
    -                                    revert(_42, _3)
    -                                }
    -                                let var_token1_1 := shr(96, calldataload(add(expr_component_2, 46)))
    -                                let expr_5 := lt(var_token1_1, var_token1)
    -                                let _43 := and(fun_computePoolAddress(var_token1, var_token1_1, and(shr(72, _38), _36)), _33)
    -                                let expr_6 := 0
    -                                switch expr_5
    -                                case 0 {
    -                                    expr_6 := 0xfffd8963efd1fc6a506488495d951d5263988d25
    -                                }
    -                                default { expr_6 := 0x01000276a4 }
    -                                let expr_mpos := mload(_2)
    -                                mstore(add(expr_mpos, 32), _2)
    -                                let tail := abi_encode_bytes_calldata(_37, lengthOut, add(expr_mpos, 96))
    -                                mstore(add(expr_mpos, _2), _34)
    -                                let _44 := sub(tail, expr_mpos)
    -                                mstore(expr_mpos, add(_44, not(31)))
    -                                finalize_allocation(expr_mpos, _44)
    -                                let _45 := mload(_2)
    -                                mstore(_45, shl(227, 0x02515961))
    -                                mstore(add(_45, _3), caller())
    -                                mstore(add(_45, 36), expr_5)
    -                                mstore(add(_45, 68), _41)
    -                                mstore(add(_45, 100), and(expr_6, _33))
    -                                mstore(add(_45, 132), 160)
    -                                let _46 := call(gas(), _43, 0, _45, sub(abi_encode_bytes(expr_mpos, add(_45, 164)), _45), _45, _2)
    -                                if iszero(_46)
    -                                {
    -                                    let pos_2 := mload(_2)
    -                                    returndatacopy(pos_2, 0, returndatasize())
    -                                    revert(pos_2, returndatasize())
    -                                }
    -                                if _46
    -                                {
    -                                    let _47 := _2
    -                                    if gt(_2, returndatasize()) { _47 := returndatasize() }
    -                                    finalize_allocation(_45, _47)
    -                                    if slt(sub(add(_45, _47), _45), _2) { revert(0, 0) }
    -                                }
    +                                let expr_component_5, expr_component_6, expr_component_7 := fun_swap(negate_int256(expr_component_4), caller(), _36, add(expr_component_1, not(22)), _34)
                                 }
    ```

### Repro on the CLI
Here's my CLI repro. For the ease of debugging it skips some options that are not relevant and also uses the same input files in both cases, just changing the command.

```bash
json_file="input1.json"
for source_unit_name in $(jq --raw-output '.sources | keys[]' "$json_file"); do
    mkdir -p ./"$(dirname "$source_unit_name")"
    jq --raw-output '.sources["'"$source_unit_name"'"].content' "$json_file" > ./"$source_unit_name"
done
jq '.settings' "$json_file" --indent 4
```

```bash
function solc-universal-router {
    solc \
        --via-ir \
        --optimize \
        --no-cbor-metadata \
        --debug-info none \
        "solmate/=lib/solmate/" \
        "permit2/=lib/permit2/" \
        "forge-std/=lib/forge-std/src/" \
        "@openzeppelin/=node_modules/@openzeppelin/" \
        "@uniswap/=node_modules/@uniswap/" \
        "ds-test/=lib/forge-std/lib/ds-test/src/" \
        "$@"
}

function solc-universal-router-bin {
    solc-universal-router --bin "$@" |
        grep '======= contracts/UniversalRouter.sol:UniversalRouter =======' --after-context=2
}

diff --color --unified --report-identical-files \
    <(solc-universal-router-bin contracts/UniversalRouter.sol contracts/DummyContract.sol | fold --width 100) \
    <(solc-universal-router-bin contracts/UniversalRouter.sol                             | fold --width 100)

diff --color --unified --report-identical-files \
    <(solc-universal-router --ir-optimized contracts/UniversalRouter.sol contracts/DummyContract.sol) \
    <(solc-universal-router --ir-optimized contracts/UniversalRouter.sol)
```

**Comment by kuzdogan** [comment]
I've come across a similar issue at https://github.com/foundry-rs/foundry/issues/8107 which even has the exact same contract (with `v0.8.17`)

However I'd have expected the issue to be resolved once I had removed the "extra" source file but I still got differences in the bytecode. Is this expected?

These are the two json files with only difference in the extra source file `lib/permit2/src/interfaces/IEIP712.sol`:

[universalrouter-sourcify-input.json.txt](https://github.com/user-attachments/files/15742204/universalrouter-sourcify-input.json.txt)
[std-json.json.txt](https://github.com/user-attachments/files/15742219/std-json.json.txt)


<details>
<summary>Script to compare `sources` path names between Foundry's standard JSON input and Sourcify's</summary>

```js
const fs = require("fs");

// Function to read and parse JSON files
function readJSONFile(filePath) {
  try {
    const data = fs.readFileSync(filePath, "utf8");
    return JSON.parse(data);
  } catch (error) {
    console.error(`Error reading or parsing file ${filePath}:`, error);
    process.exit(1);
  }
}

// Get file paths from command-line arguments
const [file1Path, file2Path] = process.argv.slice(2);

if (!file1Path || !file2Path) {
  console.error("Please provide paths for both files as arguments.");
  process.exit(1);
}

// Read and parse the JSON files
const file1 = readJSONFile(file1Path);
const file2 = readJSONFile(file2Path);

// Extract the keys of the sources objects
const file1Keys = Object.keys(file1.sources);
const file2Keys = Object.keys(file2.sources);

// Find keys that are in file1 but not in file2
const inFile1NotInFile2 = file1Keys.filter((key) => !file2Keys.includes(key));

// Find keys that are in file2 but not in file1
const inFile2NotInFile1 = file2Keys.filter((key) => !file1Keys.includes(key));

// Output the results
console.log(`Keys in ${file1Path} but not in ${file2Path}:`, inFile1NotInFile2);
console.log(`Keys in ${file2Path} but not in ${file1Path}:`, inFile2NotInFile1);
```

</details>

**Comment by nikola-matic** [comment]
On current develop? Weird, I think we fixed another one of these for the last release as well.

**Comment by cameel** [comment]
These bugs are quite insidious and we still haven't eliminated all of them. For example I recently found #15134. Unless you're absolutely sure it's the same exact bug it would be best to report it as a new issue. It's very likely that you just discovered a new but similar bug.

Also, like I commented in https://github.com/ethereum/solidity/issues/15134#issuecomment-2155411288, we decided to stop doing whack-a-mole with them and we're changing how we handle names in the optimizer and this should just wholesale fix any such errors.

Your repro will still be useful for confirming that it's actually fixed though.

**Comment by kuzdogan** [comment]
To be clear the issue was on v0.8.17 and not latest. The contract did not have the consistency issue with 0.8.26. I just wanted to point out **for v0.8.17**  that I expected the issue to not happen after removing the "extra" source file, but there were still differences. I'm not sure if that's indeed the case as I don't have a full understanding. Wanted to share in case it helps somehow

**Comment by nikola-matic** [comment]
> To be clear the issue was on v0.8.17 and not latest. The contract did not have the consistency issue with 0.8.26. I just wanted to point out **for v0.8.17**  that I expected the issue to not happen after removing the "extra" source file, but there were still differences. I'm not sure if that's indeed the case as I don't have a full understanding. Wanted to share in case it helps somehow

Ah OK, in that case it's more likely than not that this is fixed as of the 0.8.27 (26?) release, but we'll have to check on Monday regardless. Anyway, as Kamil said, we're working on fixing the root cause of these types of bugs.

**Comment by cameel** [comment]
> Ah OK, in that case it's more likely than not that this is fixed as of the 0.8.27 (26?) release, but we'll have to check on Monday regardless.

@nikola-matic It's fixed on the current version, @kuzdogan said just that. It's only 0.8.17 that's in question. The question is really why working around it by removing the file does not work on that version.

> However I'd have expected the issue to be resolved once I had removed the "extra" source file but I still got differences in the bytecode. Is this expected?
>
> These are the two json files with only difference in the extra source file `lib/permit2/src/interfaces/IEIP712.sol`:

@kuzdogan Thanks for clarifying. Looks like I initially misunderstood your question. Now I diffed the two JSON inputs and I think the bytecode differences you're seeing are not due to this bug. It's just that the extra file isn't the only difference in the input. There are a few more:
1. Different order of remappings.
1. `settings.metadata.useLiteralContent` set to `false` (default) in one but not the other.
1. Different outputs requested.
1. Code differences in two source files (other than `IEIP712.sol`).

The first three things shouldn't affect the bytecode, but the fourth one definitely does. Especially the difference in inheritance hierarchy of `IAllowanceTransfer` and in the assembly.

### Differences in `lib/permit2/src/interfaces/IAllowanceTransfer.sol`
```diff
@@ -2 +2,3 @@
-pragma solidity ^0.8.17;
+pragma solidity ^0.8.0;
+
+import {IEIP712} from "./IEIP712.sol";
@@ -7 +9 @@
-interface IAllowanceTransfer {
+interface IAllowanceTransfer is IEIP712 {
@@ -54 +56 @@
-    /// @notice The permit message signed for a single token allownce
+    /// @notice The permit message signed for a single token allowance
@@ -109 +111,4 @@
-    function allowance(address, address, address) external view returns (uint160, uint48, uint48);
+    function allowance(address user, address token, address spender)
+        external
+        view
+        returns (uint160 amount, uint48 expiration, uint48 nonce);
```

### Differences in `lib/solmate/src/utils/SafeTransferLib.sol`
```diff
@@ -46,3 +46,3 @@
-            mstore(add(freeMemoryPointer, 4), from) // Append the "from" argument.
-            mstore(add(freeMemoryPointer, 36), to) // Append the "to" argument.
-            mstore(add(freeMemoryPointer, 68), amount) // Append the "amount" argument.
+            mstore(add(freeMemoryPointer, 4), and(from, 0xffffffffffffffffffffffffffffffffffffffff)) // Append and mask the "from" argument.
+            mstore(add(freeMemoryPointer, 36), and(to, 0xffffffffffffffffffffffffffffffffffffffff)) // Append and mask the "to" argument.
+            mstore(add(freeMemoryPointer, 68), amount) // Append the "amount" argument. Masking not required as it's a full 32 byte type.
@@ -79,2 +79,2 @@
-            mstore(add(freeMemoryPointer, 4), to) // Append the "to" argument.
-            mstore(add(freeMemoryPointer, 36), amount) // Append the "amount" argument.
+            mstore(add(freeMemoryPointer, 4), and(to, 0xffffffffffffffffffffffffffffffffffffffff)) // Append and mask the "to" argument.
+            mstore(add(freeMemoryPointer, 36), amount) // Append the "amount" argument. Masking not required as it's a full 32 byte type.
@@ -111,2 +111,2 @@
-            mstore(add(freeMemoryPointer, 4), to) // Append the "to" argument.
-            mstore(add(freeMemoryPointer, 36), amount) // Append the "amount" argument.
+            mstore(add(freeMemoryPointer, 4), and(to, 0xffffffffffffffffffffffffffffffffffffffff)) // Append and mask the "to" argument.
+            mstore(add(freeMemoryPointer, 36), amount) // Append the "amount" argument. Masking not required as it's a full 32 byte type.
```

**Comment by nikola-matic** [comment]
At this point the only useful thing about this is another repro to add to the list, since we aim to fix sll issues of this type. Regardless, it's nice to know that this *is* fixed in the latest version anyway.
